### DISCUSSION TITLE
:new only works in first source

### DISCUSSION THREAD

**Discussion by lrustand** [minad/consult#1089] [Q&A]
I'm using `consult--multi` to select from some custom buffer sources. The help text for consult--multi says that there is a source field named `:new`, which is a function that is called with a a new entry.

If I add a `:new` field to the _first_ source in the list passed to consult-multi, then it works as expected, but if I instead add a `:new` field to any of the sources that are not first in the list it is not executed, even if it is the only source that has a `:new` field.

**Comment by lrustand** [comment]
Additionally, it is not possible to add the `:new` field directly to the `consult--multi` call. This would probably be even more preferrable.

**Comment by minad** [comment]
> If I add a :new field to the first source in the list passed to consult-multi, then it works as expected, but if I instead add a :new field to any of the sources that are not first in the list it is not executed, even if it is the only source that has a :new field.

No, it also works if you *narrow* to another source with a :new field. If you don't narrow, the first source wins.

**Comment by lrustand** [comment]
But why should the first source "win" if it doesn't have anything to win with? Why can't the first source that HAS a :new field win?

**Reply by minad** [reply]
I find it more consistent like this. The candidates of the first source are also matched first. A source with a `:default` field is also preferred. See the following code for the order of preference:

https://github.com/minad/consult/blob/e102b102b779355574d10de80c87603ea0086612/consult.el#L2787-L2794

You are right that we could prefer a source with a `:new` field, but then we could get an inconsistency between `:default` and `:new`.

> ...if it doesn't have anything to win with?

The question is which source the candidate will be assigned to - see the return value of `consult--multi`. The presence of a `:new` field is orthogonal and only makes it more convenient to add an additional action to execute for new candidates.
